Code/GraphMol/Depictor/Templates.h

```diff
@@ -69,11 +72,147 @@ class RDKIT_DEPICTOR_EXPORT CoordinateTemplates {
+    // mean it also intersects the vector 1. Vector 1 is only a subset of that
+    // infinite line 1, so it may have intersected that line before the vector
+    // started or after it ended. To know for sure, we have to repeat the
+    // the same test the other way round. We start by calculating the
+    // infinite line 2 in linear equation standard form.
+    a2 = v2y2 - v2y1;
+    b2 = v2x1 - v2x2;
+    c2 = (v2x2 * v2y1) - (v2x1 * v2y2);
+
+    // Calculate d1 and d2 again, this time using points of vector 1.
+    d1 = (a2 * v1x1) + (b2 * v1y1) + c2;
+    d2 = (a2 * v1x2) + (b2 * v1y2) + c2;
+
+    // Again, if both have the same sign (and neither one is 0),
+    // no intersection is possible.
+    if (d1 > 0 && d2 > 0) return false;
+    if (d1 < 0 && d2 < 0) return false;
+
+    // If they are not collinear, they must intersect in exactly one point.
+    return true;
+  }
+
+  static bool pointInsideRing(const RDGeom::Point2D &pt,
+                              const RDGeom::POINT3D_VECT &coords,
+                              const RDKit::INT_VECT &ring) {
+    // check if a point is inside a ring
+    // bounding box check
+    RDGeom::Point2D minPt(1e8, 1e8);
+    RDGeom::Point2D maxPt(-1e8, -1e8);
+    for (auto atom : ring) {
+      const auto &loc = coords.at(atom);
+      minPt.x = std::min(minPt.x, loc.x);
+      minPt.y = std::min(minPt.y, loc.y);
+      maxPt.x = std::max(maxPt.x, loc.x);
+      maxPt.y = std::max(maxPt.y, loc.y);
+    }
+    if (pt.x < minPt.x || pt.x > maxPt.x || pt.y < minPt.y || pt.y > maxPt.y) {
+      return false;
+    }
+    // ray casting : check how many times a ray to the point crosses the ring.
+    // If it crosses an odd number of times, the point is inside the ring
+    float outsideX = minPt.x - 0.1;
+    int intersections = 0;
+    for (size_t i = 0; i < ring.size(); ++i) {
+      RDGeom::Point2D p1 = coords.at(ring[i]);
+      RDGeom::Point2D p2 = coords.at(ring[(i + 1) % ring.size()]);
+      if (crossing(p1, p2, pt, RDGeom::Point2D(outsideX, pt.y)) &&
+          pt.y != p2.y) {
+        ++intersections;
+      }
+    }
+    return (intersections % 2) == 1;
+  }
+
+  static void prepareTemplate(std::shared_ptr<RDKit::RWMol> &mol) {
+    // set all atoms to C to avoid matching by element
+    for (auto &atom : mol->atoms()) {
+      atom->setAtomicNum(6);
+    }
+    boost::dynamic_bitset<> inner_atoms(mol->getNumAtoms());
+    for (auto &atom : mol->atoms()) {
+      auto point = RDGeom::Point3D(0, 0, 0);
+      // average the position of substituents
+      if (atom->getDegree() < 1) {
+        continue;
+      }
+      auto coordinates = mol->getConformer().getAtomPos(atom->getIdx());
+      for (const auto &nbrIdx :
+           boost::make_iterator_range(mol->getAtomNeighbors(atom))) {
+        point += mol->getConformer().getAtomPos(nbrIdx);
+      }
+      point /= mol->getAtomDegree(atom);
+      point = coordinates + (point - coordinates) * (-0.3);
```

Comment:
note to self: the length here is dependent on the angle of substituents... should probably normalize it. Not a big deal since we only need it to be "short"